https://issues.redhat.com/browse/ACM-25181 (#8339)
open-cluster-management/rhacm-docs: gitops/gitops_addon_with_argocd.adoc

[#enable-gitops-addon-with-argocd]

> = Enabling {gitops} add-on with _ArgoCD_ agent

Reviewer: Feedback for the entire collection here:

This was a lot, great job. I think it's getting close to done.

Themes:

- The YAML additions are missing details or specifics.

- All of them are also missing the step to apply the command. I have shown this previously, but there are also examples in the doc if you need to see what I mean. We give them a sample to add, explain it a bit, then give them the command to apply.

**Usually something like:**

1. Create the YAML file with the ___ specifications. (or resource, or label, or annotation--depends on what it actually is)
2. Apply the YAML file with the following command:
`oc apply` (check for sure)

- There are a few space issues, I may not have caught all of them.
- One file needs to be broken out into three
- Remove callouts because we cannot create those anymore, unfortunately. We all agreed to move forward with more conversion friendly content as much as we can.

Couple other smaller things.

Reviewer: I tried to walk myself through the first few steps and it's not very clear. See comments.

Author: @mikeshng and @fmt47 please see this comment. I addressed Brandi's concern in some instances. 

Looking at some of our steps, if I understand the technology write, we don't have an `oc` command for users to run to apply the YAML. Instead, the point of the action is to "create a gitops cluster resource" by applying the YAML we provide. 

I left comments at these instances. 

Please let me know if there are more steps for me to include, like what Brandi suggested here, thanks!

Reviewer 2: I am ok with either:

a)
every yaml sample to have a file name and then we say you apply the changes by
`oc apply -f the_sample_file_name.yaml`

b)
On the hub cluster, edit and apply the following yaml :
`some yaml content here`

> Enable the {gitops-short} add-on for `Advanced` pull model with the `ArgoCD` agent to  get detailed statuses on the health of your hub clusters and to simplify your cluster management processes. To enable a GitOps add-on with the `ArgoCD` agent, complete the following sections: 

Reviewer: Why do we have Advanced coded here?

Reviewer: missing comma spacing issue.

> Enable the {gitops-short} add-on for `Advanced` pull model with the `ArgoCD` agent to  get detailed statuses on the health of your hub clusters and to simplify your cluster management processes. To enable a GitOps add-on with the `ArgoCD` agent, complete the following sections: 

Reviewer: ```suggestion
Enable the {gitops-short} add-on for `Advanced` pull model with the `ArgoCD` agent to get detailed statuses on the health of your hub clusters, and to simplify your cluster management processes. To enable a GitOps add-on with the `ArgoCD` agent, complete the following sections: 
```

> Enable the {gitops-short} add-on for `Advanced` pull model with the `ArgoCD` agent to  get detailed statuses on the health of your hub clusters and to simplify your cluster management processes. To enable a GitOps add-on with the `ArgoCD` agent, complete the following sections: 

Reviewer 2: ```suggestion
Enable the {gitops-short} add-on for `Advanced` pull model with the `ArgoCD` agent to get detailed statuses on the health of your Application. To enable a GitOps add-on with the `ArgoCD` agent, complete the following sections: 
```

* <<prerequisites, Prerequisites>>
* <<enable-argocd-agent, Enabling the _ArgoCD_ agent>>
* <<verify-installation, Verifying the installation>>
* <<additional-resources, Additional resources>> 

[#prerequisites]
== Prerequisites 

> * {acm-short} hub cluster installed

Reviewer: Throughout the doc, these are complete sentences. We would be changing that with this prereq list. Usually:

You need x 

or 

You have X.

* Managed clusters registered with {acm-short}

> * OpenShift GitOps operator installed on the hub cluster

Reviewer: I think this is conref-ed in our docs--please check; please check on all product names internally and check the guidance if we don't have anything for this. Check all OCP and GitOps instances for conref and naming.

* A `Placement` resource to select target managed clusters
* `ManagedClusterSet` bound to the target namespace
* OpenShift GitOps operator subscription configured with the `ArgoCD` agent environment 
* The `ArgoCD` custom resource configured for the `Agent` mode

[#configure-subscriptions-resources]
== Configuring subscriptions and resources 

To enable the `ArgoCD` agent, you must configure the {gitops-short} operator subscription and the `ArgoCD` custom resource. To configure the necessary subscriptions and resources, complete the following steps:

. On the hub cluster only, modify the OpenShift GitOps operator subscription to include the required environment variables:

+
[source,bash]
----
oc edit subscription gitops-operator -n openshift-gitops-operator
----

> . Add the following environment variables under `spec.config.env`:

Reviewer: please never use under, this can alienate a nonsighted user. We go over this in our internal doc very clearly: https://stolostron.github.io/rhacm-docs/accessibility

Further we have this on the checklist: https://stolostron.github.io/rhacm-docs/checklist (directional language)

+
[source,yaml]
----
spec:
  config:
    env:
    - name: ARGOCD_CLUSTER_CONFIG_NAMESPACES
      value: openshift-gitops
    - name: ARGOCD_PRINCIPAL_TLS_SERVER_ALLOW_GENERATE
      value: "false"
    - name: ARGOCD_PRINCIPAL_REDIS_SERVER_ADDRESS
      value: openshift-gitops-redis:6379
----

. Replace the existing `ArgoCD` custom resource with the compatible `Agent` mode configuration:

+
[source,yaml]
----
apiVersion: argoproj.io/v1beta1
kind: ArgoCD
metadata:
  name: openshift-gitops
  namespace: openshift-gitops
spec:
  controller:
    enabled: false
  argoCDAgent:
    principal:
      allowedNamespaces:
      - '*'
      auth: mtls:CN=system:open-cluster-management:cluster:([^:]+):addon:gitops-addon:agent:gitops-addon-agent
      enabled: true
----

* *Note:* On the hub cluster only, this configuration disables the traditional `ArgoCD` controller and enables the agent principal with mutual TLS authentication.

[#enable-argocd-agent]
== Enabling _ArgoCD_ agent 

> Create a `GitOpsCluster` resource to manage the `ArgoCD` agent add-on deployment. its controller automatically creates the following resources for each managed cluster selected by the `Placement`

Reviewer: Remove the possessive here and rewrite that line; see the IBM style guide about possessive.

"its" is also not capitalized, but is a start a new sentence, I think?

The `GitOpsCluster` controller performs the following operations:

> * Create and automated PKI management 

Reviewer: The controller .... creates (add the s to make subject/predicate agree,

> * Creates `ArgoCD` cluster secrets configured for agent mode

Reviewer: that are?

Reviewer 2: They are for Argo CD to understand what clusters are imported. 
Argo CD by default does not understand ACM's ManagedClusters. 
This progress is equivalent of: importing all the ACM's ManagedClusters (selected by the Placement) into Argo CD.

> * Deploys the Argo CD Agent on each selected managed cluster

Reviewer: Sometimes Argo CD is coded, see above examples, and sometimes not.

> To create a `GitOpsCluster` resource with the `ArgoCD` agent, complete the following steps:

Reviewer 2: ```suggestion
To enable the advanced pull model Argo CD Agent architecture, complete the following steps:
```
Does that sound better? Or remove the whole line because the next line's context sounds good to me but this current one repeats it.

> . Create a `GitOpsCluster resource with the `ArgoCD` agent enabled by inputting the following YAML sample: 

Reviewer: inputting? 

We normally keep it simple:

Add the following YAML to create ___

Create ___ by adding the following YAML.

Are they adding the entire sample and what from it matters to the user. We don't want to just give them a big sample and not explain it.

+
[source,yaml]
----
apiVersion: apps.open-cluster-management.io/v1beta1
kind: GitOpsCluster
metadata:
  name: gitops-agent-clusters
  namespace: openshift-gitops
spec:
  argoServer:
    argoNamespace: openshift-gitops
  placementRef:
    kind: Placement
    apiVersion: cluster.open-cluster-management.io/v1beta1
    name: production-clusters
    namespace: openshift-gitops
  gitopsAddon:
    enabled: true
    argoCDAgent:
      enabled: true

>       propagateHubCA: true

Reviewer 2: ```suggestion
```

Let's not show this field that is default to true all the time and doesn't need to be manually set and the user will rarely ever use to set it to false.

----

[#verify-installation]
== Verifying the installation 

> After the `ArgoCD` agent is successfully deployed, verify the `Pull Model` workflow is completed by creating an application on the hub cluster and confirming it works on the managed cluster. 

Reviewer 2: ```suggestion
After the `ArgoCD` agent is successfully deployed, verify the advanced `Pull Model` workflow is completed by creating an application on the hub cluster and confirming it works on the managed cluster. 
```

I like your idea of advanced pull model vs basic pull model so I am stealing that usage here.

To verify the necessary installations and resources for successful deployment, complete the following steps:

. Check the `GitOpsCluster` status for specific `Agent` conditions by running the following command:

+
[source,bash]
----
oc get gitopscluster gitops-agent-clusters -n openshift-gitops -o jsonpath='{.status.conditions}' | jq
----

. Confirm that you see the following condition types in the status:

+
* `ArgoCDAgentPrereqsReady` - Agent prerequisites are set up
* `CertificatesReady` - TLS certificates are signed

> * `ManifestWorksApplied` - CA certificates propagated to managed clusters

Reviewer 2: - Ready - GitOpsCluster is ready and all components are functioning correctly
- PlacementResolved - Placement reference resolved and managed clusters retrieved
- ClustersRegistered - Managed clusters successfully registered with ArgoCD server
- AddOnDeploymentConfigsReady - AddOnDeploymentConfigs created for all managed clusters (GitOps addon)
- ManagedClusterAddOnsReady - ManagedClusterAddons created/updated for all managed clusters (GitOps addon)
- AddOnTemplateReady - Dynamic AddOnTemplate created for ArgoCD agent mode (ArgoCD agent only)
- ArgoCDAgentPrereqsReady - ArgoCD agent prerequisites are set up (RBAC, Redis secret, JWT secret, server discovery)
- CertificatesReady - ArgoCD agent TLS certificates (CA, Principal, Resource Proxy) are signed and ready
- ManifestWorksApplied - CA certificates propagated to managed clusters via ManifestWorks (ArgoCD agent only)

Sorry, we added new conditions. Please see above for the full list.

. On the hub cluster, create an `ArgoCD` application resource in the managed cluster namespace by inputting the following YAML file: 

+
[source,yaml]
----
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: guestbook
  namespace: <managed cluster name>
spec:
  project: default
  source:
    repoURL: https://github.com/argoproj/argocd-example-apps.git
    targetRevision: HEAD
    path: guestbook
  destination:
    server: https://<principal-external-ip:port>?agentName=<managed cluster name>
    namespace: guestbook
  syncPolicy:
    automated:
      prune: true
      selfHeal: true
----

. On the managed cluster, verify that the application resources are deployed by running the following command:

+
[source,bash]
----
oc get all -n guestbook
----

. On the hub cluster, verify the application status is reflected back to you by running the following command:

+
[source,bash]
----
oc get application guestbook -n <managed cluster name>
----

. Confirm that the status shows  `Synced` when the application is successfully deployed.

[#additional-resources]
== Additional resources 

> Continue learning about the {gitops-short} add-on by completing xref:../gitops/gitops_manage_addon.adoc#manage-gitops-addon[Managing the {gitops} add-on].

Reviewer: I would take out learning here since they are heading to a task and not a concept and you are telling them to "complete" something?

* <<verify-installation, Verifying the installation>>
* <<additional-resources, Additional resources>> 

[#prerequisites]
== Prerequisites 

Ensure that you have the following resources installed: 

* {acm-short} hub cluster installed
* Managed clusters registered with {acm-short}
* {gitops-short} operator installed on the hub cluster
* A `Placement` resource to select target managed clusters
* `ManagedClusterSet` bound to the target namespace
* {gitops-short} operator subscription configured with the `ArgoCD` agent environment 
* The `ArgoCD` custom resource that are configured for the `Agent` mode

[#configure-subscriptions-resources]
== Configuring subscriptions and resources 

To enable the `ArgoCD` agent, you must configure the {gitops-short} operator subscription and the `ArgoCD` custom resource. To configure the necessary subscriptions and resources, complete the following steps:

. On the hub cluster only, modify the {gitops-short} operator subscription to include the required environment variables by running the following command:

+
[source,bash]
----
oc edit subscription gitops-operator -n openshift-gitops-operator
----

> . Add the following environment variables to `spec.config.env` by applying the following YAML sample:

Reviewer: to the  to `spec.config.env` file? This needs some clarity.

Reviewer: Then we need to say what it is that is important from this YAML.

Typically it needs to be something like this:

Create a YAML file with the following x resource to (action here).

Run the following command.

I am not sure where this came from or what it is, are they given this:

`spec.config.env`

* <<verify-installation, Verifying the installation>>
* <<additional-resources, Additional resources>> 

[#prerequisites]
== Prerequisites 

Ensure that you have the following resources installed: 

* {acm-short} hub cluster installed
* Managed clusters registered with {acm-short}
* {gitops-short} operator installed on the hub cluster
* A `Placement` resource to select target managed clusters
* `ManagedClusterSet` bound to the target namespace
* {gitops-short} operator subscription configured with the `ArgoCD` agent environment 
* The `ArgoCD` custom resource that are configured for the `Agent` mode

[#configure-subscriptions-resources]
== Configuring subscriptions and resources 

To enable the `ArgoCD` agent, you must configure the {gitops-short} operator subscription and the `ArgoCD` custom resource. To configure the necessary subscriptions and resources, complete the following steps:

. On the hub cluster only, modify the {gitops-short} operator subscription to include the required environment variables by running the following command:

+
[source,bash]
----
oc edit subscription gitops-operator -n openshift-gitops-operator
----

> . Add the following environment variables to the `spec.config.env` file by applying the following YAML sample:

Reviewer: `to the `spec.config.env` file`... this is good if it is a file we are referencing. 

I would just add:

Save the file. (Making it truly a step-by-step procedure.)

+
[source,yaml]
----
apiVersion: argoproj.io/v1beta1
kind: ArgoCD
metadata:
  name: openshift-gitops
  namespace: openshift-gitops
spec:
  controller:
    enabled: false
  argoCDAgent:
    principal:
      allowedNamespaces:
      - '*'
      auth: mtls:CN=system:open-cluster-management:cluster:([^:]+):addon:gitops-addon:agent:gitops-addon-agent
      enabled: true
----

* *Note:* On the hub cluster only, this configuration disables the traditional `ArgoCD` controller and enables the agent principal with mutual TLS authentication.

[#enable-argocd-agent]
== Enabling _ArgoCD_ agent 

Create a `GitOpsCluster` resource to manage the `ArgoCD` agent add-on deployment. The controller controller automatically creates the following resources for each managed cluster selected by the `Placement`

The `GitOpsCluster` controller performs the following operations:

* Creates and automated PKI management 
* Creates `ArgoCD` cluster secrets configured for agent mode
* Deploys the Argo CD Agent on each selected managed cluster

To enable the advanced pull model Argo CD Agent architecture, complete the following steps:

> . On your managed cluster, create a `GitOpsCluster resource with the `ArgoCD` agent enabled by applying the following YAML sample: 

Author: @mikeshng and/ or @fmt47 , like Brandi suggested, she wants details on where users are running these commands. I see for the steps on "configuring subscriptions" its on the hub cluster only, so I thought these steps were on the managed cluster. 

Please correct me if I am wrong.

Author: Also, I'll fix the highlighting issue in a new commit when I address more feedback. Thanks!

Reviewer: This is usually a simple command to apply @jc-berger -- we have some examples in the doc.

> . On your managed cluster, create a `GitOpsCluster resource with the `ArgoCD` agent enabled by applying the following YAML sample: 

Reviewer: This is off, I think, meaning a whole phrase is coded? Do you mean to only code `GitOpsCluster` (seems so with the YAML)

`GitOpsCluster resource with the `ArgoCD` is off.

> . On your managed cluster, create a `GitOpsCluster resource with the `ArgoCD` agent enabled by applying the following YAML sample: 

Reviewer 2: ```suggestion
. On your hub cluster, create a `GitOpsCluster resource with the `ArgoCD` agent enabled by applying the following YAML sample: 
```

GitOpsCluster CR should be created in the hub cluster.

+
[source,yaml]
----
apiVersion: apps.open-cluster-management.io/v1beta1
kind: GitOpsCluster
metadata:
  name: gitops-agent-clusters
  namespace: openshift-gitops
spec:
  argoServer:
    argoNamespace: openshift-gitops
  placementRef:
    kind: Placement
    apiVersion: cluster.open-cluster-management.io/v1beta1
    name: production-clusters
    namespace: openshift-gitops
  gitopsAddon:
    enabled: true
    argoCDAgent:
      enabled: true

> ----

Reviewer: Still missing anything that tells the user how to apply the change.

+
[source,bash]
----
oc get gitopscluster gitops-agent-clusters -n openshift-gitops -o jsonpath='{.status.conditions}' | jq
----

. Confirm that you see the following condition types in the status:

+
* `Ready`: The `GitOpsCluster` is ready and all the components are functioning. 
* `PlacementResolved`: The `Placement` reference is resolved and the managed clusters are retrieved. 
* `ClustersRegistered`: The managed clusters are successfully registered with  the `ArgoCD` server.
* `AddOnDeploymentConfigsReady`: The `AddOnDeploymentConfigs` are created for all the managed clusters.
* `ManagedClusterAddOnsReady`: The `ManagedClusterAddons` creates and updates the managed clusters. 
* `AddOnTemplateReady`: The dynamic `AddOnTemplate` created for the `ArgoCD` agent mode. 
* `ArgoCDAgentPrereqsReady`: The agent prerequisites are set up.
* `CertificatesReady`: The TLS certificates are signed.
* `ManifestWorksApplied`: The CA certificates propagated to managed clusters.

. On the hub cluster, create an `ArgoCD` application resource in the managed cluster namespace by inputting the following YAML file: 

> 

Reviewer: just _adding_ or _applying_ instead of _inputting_

+
[source,yaml]
----
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: guestbook
  namespace: <managed cluster name>
spec:
  project: default
  source:
    repoURL: https://github.com/argoproj/argocd-example-apps.git
    targetRevision: HEAD
    path: guestbook
  destination:
    server: https://<principal-external-ip:port>?agentName=<managed cluster name>
    namespace: guestbook
  syncPolicy:
    automated:
      prune: true
      selfHeal: true

> ----

Reviewer: missing how to apply

+
[source,bash]
----
oc edit subscription gitops-operator -n openshift-gitops-operator
----

. Add the following environment variables to the `spec.config.env` file by applying the following YAML sample:

+
[source,yaml]
----
spec:
  config:
    env:
    - name: ARGOCD_CLUSTER_CONFIG_NAMESPACES
      value: openshift-gitops
    - name: ARGOCD_PRINCIPAL_TLS_SERVER_ALLOW_GENERATE
      value: "false"
    - name: ARGOCD_PRINCIPAL_REDIS_SERVER_ADDRESS
      value: openshift-gitops-redis:6379
----

. Replace the existing `ArgoCD` custom resource with the compatible `Agent` mode configuration by applying the following YAML sample:

+
[source,yaml]
----
apiVersion: argoproj.io/v1beta1
kind: ArgoCD
metadata:
  name: openshift-gitops
  namespace: openshift-gitops
spec:
  controller:
    enabled: false
  argoCDAgent:
    principal:
      allowedNamespaces:
      - '*'
      auth: mtls:CN=system:open-cluster-management:cluster:([^:]+):addon:gitops-addon:agent:gitops-addon-agent
      enabled: true
----

* *Note:* On the hub cluster only, this configuration disables the traditional `ArgoCD` controller and enables the agent principal with mutual TLS authentication.

> . Finish applying and save the YAML sample that you edited by running the following command:  

Author: @mikeshng and @fmt47 does this structure work? 

Basically, at the end of the procedure, we tell users to finish applying and saving the yaml they've been editing? 

Or would we need to provide this command after every YAML sample we give? From our discussions, it seems like this command is a formal way to save and apply all the changes, so I think it makes sense to put at the end, but please let me know what you think :) Thanks!

Reviewer 2: Putting it at the end makes sense but then every yaml sample needs to have a sample name or maybe something like `oc apply -f yaml_file_name`

* Managed clusters registered with {acm-short}
* {gitops-short} operator installed on the hub cluster
* A `Placement` resource to select target managed clusters
* `ManagedClusterSet` bound to the target namespace
* {gitops-short} operator subscription configured with the `ArgoCD` agent environment 
* The `ArgoCD` custom resource that are configured for the `Agent` mode

To enable the {gitops-short} add-on with the `ArgoCD` agent, complete the following sections: 

* <<enable-argocd-agent, Enabling the _ArgoCD_ agent>>
* <<verify-installation, Verifying the installation>>
* <<additional-resources, Additional resources>> 

[#configure-subscriptions-resources]
== Configuring subscriptions and resources 

To enable the `ArgoCD` agent, you must configure the {gitops-short} operator subscription and the `ArgoCD` custom resource. To configure the necessary subscriptions and resources, complete the following steps:

. On the hub cluster only, modify the {gitops-short} operator subscription to include the required environment variables by running the following command:

+
[source,bash]
----
oc edit subscription gitops-operator -n openshift-gitops-operator
----

> . Add the following environment variables to the `spec.config.env` file by applying the following YAML sample:

Reviewer: ```suggestion
. Add the environment variables to the `spec.config.env` file, as displayed in the following YAML sample:
```

+
[source,yaml]
----
spec:
  config:
    env:
    - name: ARGOCD_CLUSTER_CONFIG_NAMESPACES
      value: openshift-gitops
    - name: ARGOCD_PRINCIPAL_TLS_SERVER_ALLOW_GENERATE
      value: "false"
    - name: ARGOCD_PRINCIPAL_REDIS_SERVER_ADDRESS
      value: openshift-gitops-redis:6379
----

. Replace the existing `ArgoCD` custom resource with the compatible `Agent` mode configuration by applying the following YAML sample:

+
[source,yaml]
----
apiVersion: argoproj.io/v1beta1
kind: ArgoCD
metadata:
  name: openshift-gitops
  namespace: openshift-gitops
spec:
  controller:
    enabled: false
  argoCDAgent:
    principal:
      allowedNamespaces:
      - '*'
      auth: mtls:CN=system:open-cluster-management:cluster:([^:]+):addon:gitops-addon:agent:gitops-addon-agent
      enabled: true
----

> . Finish applying and save the YAML sample that you edited by running the following command:  

Reviewer: ```suggestion
. Run the following command to apply your YAML sample:
```

[#enable-argocd-agent]
== Enabling _ArgoCD_ agent 

Create a `GitOpsCluster` resource to manage the `ArgoCD` agent add-on deployment. The controller controller automatically creates the following resources for each managed cluster selected by the `Placement`

The `GitOpsCluster` controller performs the following operations:

* Creates and automated PKI management 
* Creates `ArgoCD` cluster secrets configured for agent mode
* Deploys the Argo CD Agent on each selected managed cluster

To enable the advanced pull model Argo CD Agent architecture, complete the following steps:

. On your managed cluster, create a `GitOpsCluster`` resource with the `ArgoCD` agent enabled by applying the following YAML sample: 

+
[source,yaml]
----
apiVersion: apps.open-cluster-management.io/v1beta1
kind: GitOpsCluster
metadata:
  name: gitops-agent-clusters
  namespace: openshift-gitops
spec:
  argoServer:
    argoNamespace: openshift-gitops
  placementRef:
    kind: Placement
    apiVersion: cluster.open-cluster-management.io/v1beta1
    name: production-clusters
    namespace: openshift-gitops
  gitopsAddon:
    enabled: true
    argoCDAgent:
      enabled: true
----

> . Finish applying and save the YAML sample by running the following command:  

Reviewer: ```suggestion
.  Run the following command to apply your YAML sample: 
```
